File: py/wlearn/scalers.py
"""Numeric preprocessing transformers (v1: numpy arrays only).

StandardScaler and MinMaxScaler mirror JS @wlearn/core implementations.
Cross-language bundle round-trips are supported: JS-saved bundles can be
loaded in Python and vice versa.
"""

import json

import numpy as np

from .errors import NotFittedError, DisposedError, ValidationError
from .bundle import encode_bundle
from .registry import register

STANDARD_SCALER_TYPE_ID = 'wlearn.preprocess.standard_scaler@1'
MINMAX_SCALER_TYPE_ID = 'wlearn.preprocess.minmax_scaler@1'


class StandardScaler:
    """Standardizes features by removing the mean and scaling to unit variance.

    Uses Welford's algorithm for numerical stability (matching JS implementation).
    Stores sample std (ddof=1) to match JS behavior.
    """

    def __init__(self, params=None):
        self._params = dict(params) if params else {}
        self._means = None
        self._stds = None
        self._fitted = False
        self._disposed = False

    @classmethod
    def create(cls, params=None):
        return cls(params)

    def fit(self, X, y=None):
        self._ensure_alive()
        X = np.asarray(X, dtype=np.float64)
        if X.ndim == 1:
            X = X.reshape(1, -1)
        if X.shape[0] == 0:
            raise ValidationError('Cannot fit on empty data')

        self._means = X.mean(axis=0)
        # Sample std (ddof=1) matching JS Welford's (rows-1 denominator)
        self._stds = X.std(axis=0, ddof=1) if X.shape[0] > 1 else np.zeros(X.shape[1])
        self._fitted = True
        return self

    def transform(self, X):
        self._ensure_fitted()
        X = np.asarray(X, dtype=np.float64)
        if X.ndim == 1:
            X = X.reshape(1, -1)
        if X.shape[1] != len(self._means):
            raise ValidationError(
                f'Expected {len(self._means)} columns, got {X.shape[1]}')
        stds = self._stds.copy()
        stds[stds == 0] = 1.0  # avoid division by zero, output 0
        result = (X - self._means) / stds
        # Where std was 0, output should be 0 (matching JS behavior)
        result[:, self._stds == 0] = 0.0
        return result

    def fit_transform(self, X, y=None):
        self.fit(X, y)
        return self.transform(X)

    def save(self):
        self._ensure_fitted()
        artifact = json.dumps(
            {'means': self._means.tolist(), 'stds': self._stds.tolist()},
            sort_keys=True, separators=(',', ':'),
        ).encode()
        return encode_bundle(
            {'typeId': STANDARD_SCALER_TYPE_ID, 'params': self.get_params()},
            [{'id': 'params', 'data': artifact, 'mediaType': 'application/json'}],
        )

    @staticmethod
    def _from_bundle(manifest, toc, blobs):
        entry = next((e for e in toc if e['id'] == 'params'), None)
        if entry is None:
            raise ValidationError('Bundle missing "params" artifact')
        data = json.loads(bytes(blobs[entry['offset']:entry['offset'] + entry['length']]))
        scaler = StandardScaler(manifest.get('params'))
        scaler._means = np.array(data['means'], dtype=np.float64)
        scaler._stds = np.array(data['stds'], dtype=np.float64)
        scaler._fitted = True
        return scaler

    def dispose(self):
        if self._disposed:
            return
        self._disposed = True
        self._means = None
        self._stds = None
        self._fitted = False

    def get_params(self):
        return dict(self._params)

    def set_params(self, p):
        self._params.update(p)
        return self

    @property
    def is_fitted(self):
        return self._fitted and not self._disposed

    def _ensure_alive(self):
        if self._disposed:
            raise DisposedError('StandardScaler has been disposed.')

    def _ensure_fitted(self):
        self._ensure_alive()
        if not self._fitted:
            raise NotFittedError('StandardScaler is not fitted.')


class MinMaxScaler:
    """Scales features to [0, 1] range based on per-column min and max."""

    def __init__(self, params=None):
        self._params = dict(params) if params else {}
        self._mins = None
        self._maxs = None
        self._fitted = False
        self._disposed = False

    @classmethod
    def create(cls, params=None):
        return cls(params)

    def fit(self, X, y=None):
        self._ensure_alive()
        X = np.asarray(X, dtype=np.float64)
        if X.ndim == 1:
            X = X.reshape(1, -1)
        if X.shape[0] == 0:
            raise ValidationError('Cannot fit on empty data')

        self._mins = X.min(axis=0)
        self._maxs = X.max(axis=0)
        self._fitted = True
        return self

    def transform(self, X):
        self._ensure_fitted()
        X = np.asarray(X, dtype=np.float64)
        if X.ndim == 1:
            X = X.reshape(1, -1)
        if X.shape[1] != len(self._mins):
            raise ValidationError(
                f'Expected {len(self._mins)} columns, got {X.shape[1]}')
        ranges = self._maxs - self._mins
        result = np.zeros_like(X)
        mask = ranges > 0
        result[:, mask] = (X[:, mask] - self._mins[mask]) / ranges[mask]
        return result

    def fit_transform(self, X, y=None):
        self.fit(X, y)
        return self.transform(X)

    def save(self):
        self._ensure_fitted()
        artifact = json.dumps(
            {'mins': self._mins.tolist(), 'maxs': self._maxs.tolist()},
            sort_keys=True, separators=(',', ':'),
        ).encode()
        return encode_bundle(
            {'typeId': MINMAX_SCALER_TYPE_ID, 'params': self.get_params()},
            [{'id': 'params', 'data': artifact, 'mediaType': 'application/json'}],
        )

    @staticmethod
    def _from_bundle(manifest, toc, blobs):
        entry = next((e for e in toc if e['id'] == 'params'), None)
        if entry is None:
            raise ValidationError('Bundle missing "params" artifact')
        data = json.loads(bytes(blobs[entry['offset']:entry['offset'] + entry['length']]))
        scaler = MinMaxScaler(manifest.get('params'))
        scaler._mins = np.array(data['mins'], dtype=np.float64)
        scaler._maxs = np.array(data['maxs'], dtype=np.float64)
        scaler._fitted = True
        return scaler

    def dispose(self):
        if self._disposed:
            return
        self._disposed = True
        self._mins = None
        self._maxs = None
        self._fitted = False

    def get_params(self):
        return dict(self._params)

    def set_params(self, p):
        self._params.update(p)
        return self

    @property
    def is_fitted(self):
        return self._fitted and not self._disposed

    def _ensure_alive(self):
        if self._disposed:
            raise DisposedError('MinMaxScaler has been disposed.')

    def _ensure_fitted(self):
        self._ensure_alive()
        if not self._fitted:
            raise NotFittedError('MinMaxScaler is not fitted.')


register(STANDARD_SCALER_TYPE_ID, StandardScaler._from_bundle)
register(MINMAX_SCALER_TYPE_ID, MinMaxScaler._from_bundle)

File: packages/core/src/scalers.js
// Numeric preprocessing transformers (v1: DenseMatrix only).
// StandardScaler and MinMaxScaler implement the Transformer interface.

const { NotFittedError, DisposedError, ValidationError } = require('./errors.js')
const { normalizeX } = require('./matrix.js')
const { encodeBundle, encodeJSON, decodeJSON } = require('./bundle.js')
const { register } = require('./registry.js')

const STANDARD_SCALER_TYPE_ID = 'wlearn.preprocess.standard_scaler@1'
const MINMAX_SCALER_TYPE_ID = 'wlearn.preprocess.minmax_scaler@1'

// --- StandardScaler ---

class StandardScaler {
  #means = null
  #stds = null
  #fitted = false
  #disposed = false
  #params = {}

  constructor(params = {}) {
    this.#params = { ...params }
  }

  fit(X) {
    this.#ensureAlive()
    const { rows, cols, data } = normalizeX(X)
    if (rows === 0) throw new ValidationError('Cannot fit on empty data')

    const means = new Float64Array(cols)
    const m2 = new Float64Array(cols)

    // Welford's online algorithm
    for (let r = 0; r < rows; r++) {
      const n = r + 1
      for (let c = 0; c < cols; c++) {
        const val = data[r * cols + c]
        const delta = val - means[c]
        means[c] += delta / n
        const delta2 = val - means[c]
        m2[c] += delta * delta2
      }
    }

    const stds = new Float64Array(cols)
    for (let c = 0; c < cols; c++) {
      stds[c] = rows > 1 ? Math.sqrt(m2[c] / (rows - 1)) : 0
    }

    this.#means = means
    this.#stds = stds
    this.#fitted = true
    return this
  }

  transform(X) {
    this.#ensureFitted()
    const { rows, cols, data } = normalizeX(X)
    if (cols !== this.#means.length) {
      throw new ValidationError(
        `Expected ${this.#means.length} columns, got ${cols}`
      )
    }

    const out = new Float64Array(rows * cols)
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const idx = r * cols + c
        const std = this.#stds[c]
        out[idx] = std > 0 ? (data[idx] - this.#means[c]) / std : 0
      }
    }
    return { rows, cols, data: out }
  }

  fitTransform(X) {
    this.fit(X)
    return this.transform(X)
  }

  save() {
    this.#ensureFitted()
    const artifact = {
      means: Array.from(this.#means),
      stds: Array.from(this.#stds),
    }
    return encodeBundle(
      { typeId: STANDARD_SCALER_TYPE_ID, params: this.getParams() },
      [{ id: 'params', data: encodeJSON(artifact), mediaType: 'application/json' }]
    )
  }

  static _fromBundle(manifest, toc, blobs) {
    const entry = toc.find(e => e.id === 'params')
    if (!entry) throw new ValidationError('Bundle missing "params" artifact')
    const artifact = decodeJSON(blobs.subarray(entry.offset, entry.offset + entry.length))
    const scaler = new StandardScaler(manifest.params || {})
    scaler.#means = new Float64Array(artifact.means)
    scaler.#stds = new Float64Array(artifact.stds)
    scaler.#fitted = true
    return scaler
  }

  dispose() {
    if (this.#disposed) return
    this.#disposed = true
    this.#means = null
    this.#stds = null
    this.#fitted = false
  }

  getParams() { return { ...this.#params } }
  setParams(p) { Object.assign(this.#params, p); return this }
  get isFitted() { return this.#fitted && !this.#disposed }

  #ensureAlive() {
    if (this.#disposed) throw new DisposedError('StandardScaler has been disposed.')
  }

  #ensureFitted() {
    this.#ensureAlive()
    if (!this.#fitted) throw new NotFittedError('StandardScaler is not fitted.')
  }
}

// --- MinMaxScaler ---

class MinMaxScaler {
  #mins = null
  #maxs = null
  #fitted = false
  #disposed = false
  #params = {}

  constructor(params = {}) {
    this.#params = { ...params }
  }

  fit(X) {
    this.#ensureAlive()
    const { rows, cols, data } = normalizeX(X)
    if (rows === 0) throw new ValidationError('Cannot fit on empty data')

    const mins = new Float64Array(cols).fill(Infinity)
    const maxs = new Float64Array(cols).fill(-Infinity)

    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const val = data[r * cols + c]
        if (val < mins[c]) mins[c] = val
        if (val > maxs[c]) maxs[c] = val
      }
    }

    this.#mins = mins
    this.#maxs = maxs
    this.#fitted = true
    return this
  }

  transform(X) {
    this.#ensureFitted()
    const { rows, cols, data } = normalizeX(X)
    if (cols !== this.#mins.length) {
      throw new ValidationError(
        `Expected ${this.#mins.length} columns, got ${cols}`
      )
    }

    const out = new Float64Array(rows * cols)
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const idx = r * cols + c
        const range = this.#maxs[c] - this.#mins[c]
        out[idx] = range > 0 ? (data[idx] - this.#mins[c]) / range : 0
      }
    }
    return { rows, cols, data: out }
  }

  fitTransform(X) {
    this.fit(X)
    return this.transform(X)
  }

  save() {
    this.#ensureFitted()
    const artifact = {
      mins: Array.from(this.#mins),
      maxs: Array.from(this.#maxs),
    }
    return encodeBundle(
      { typeId: MINMAX_SCALER_TYPE_ID, params: this.getParams() },
      [{ id: 'params', data: encodeJSON(artifact), mediaType: 'application/json' }]
    )
  }

  static _fromBundle(manifest, toc, blobs) {
    const entry = toc.find(e => e.id === 'params')
    if (!entry) throw new ValidationError('Bundle missing "params" artifact')
    const artifact = decodeJSON(blobs.subarray(entry.offset, entry.offset + entry.length))
    const scaler = new MinMaxScaler(manifest.params || {})
    scaler.#mins = new Float64Array(artifact.mins)
    scaler.#maxs = new Float64Array(artifact.maxs)
    scaler.#fitted = true
    return scaler
  }

  dispose() {
    if (this.#disposed) return
    this.#disposed = true
    this.#mins = null
    this.#maxs = null
    this.#fitted = false
  }

  getParams() { return { ...this.#params } }
  setParams(p) { Object.assign(this.#params, p); return this }
  get isFitted() { return this.#fitted && !this.#disposed }

  #ensureAlive() {
    if (this.#disposed) throw new DisposedError('MinMaxScaler has been disposed.')
  }

  #ensureFitted() {
    this.#ensureAlive()
    if (!this.#fitted) throw new NotFittedError('MinMaxScaler is not fitted.')
  }
}

// Auto-register loaders
register(STANDARD_SCALER_TYPE_ID, StandardScaler._fromBundle)
register(MINMAX_SCALER_TYPE_ID, MinMaxScaler._fromBundle)

module.exports = { StandardScaler, MinMaxScaler }
